### PR TITLE
Unify class objects into a single object per class

### PR DESCRIPTION
**Pull request by arctic5**
also removed some dead code and a small memory leak in class select controller

### CODE REVIEW

File: Source/gg2/Scripts/GameServer/processClientCommands.gml [hunk 1]
```diff
@@ -146,7 +146,7 @@ while(commandLimitRemaining > 0) {
                 
             if(balance != newTeam and newTeam != player.team)
             {
-                if(getCharacterObject(newTeam, player.class) != -1 or newTeam==TEAM_SPECTATOR)
+                if(getCharacterObject(player.class) != -1 or newTeam==TEAM_SPECTATOR)
```
Comment (Medo42) [inline Source/gg2/Scripts/GameServer/processClientCommands.gml]: This test needs fixing - you can now change to a non-existent team.

**Comment by Medo42** [comment]
Very nice, I don't see a compelling reason to keep separate objects for the team distinction. it's just that way because I thought at some point out would be "cleaner" than a switch/case. However, the way you changed it is more concise and just as readable, and actually simplifies some related code.

Apart from the issue I mentioned this seems to be good to merge.

**Comment by TazeTSchnitzel** [comment]
I wonder if this might break some mods or plugins. But it's not like GG2 has ever cared about addon compatibility

**Comment by Medo42** [comment]
If we reject all changes that possibly impact plug-in compatibility we might as well stop doing anything to the game, especially refactoring.

**Comment by TazeTSchnitzel** [comment]
Yep :/

**Comment by arctic5** [comment]
I think that should fix it

**Comment by Medo42** [comment]
While this does remedy the problem, I'd like to suggest some changes. First, using < and > with defined constants that are basically used as enums feels a bit fragile, since it's not locally obvious that these are the constants at the "limits", and which values lie between. In this case with only so few valid cases, I'd prefer just listing them all: `if(newTeam != TEAM_RED and newTeam != TEAM_BLUE and newTeam != TEAM_SPECTATOR)`

Also, the test `getCharacterObject(player.class) != -1` can be completely removed now. It used to test for a valid combination of team and class, but since the only aspect we're changing here is the team, and that no longer goes into the function at all, I think it no longer makes any real difference. Unless you can have an invalid class when you are TEAM_SPECTATOR, but I doubt that.

I'd mark this for 2.8.1, but since it's not a visible feature and might upset some mods, maybe we should wait for the next bigger update.

**Pull request merged**